zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/util/ZkPathRecursiveWatcherTrie.java

```diff
@@ -0,0 +1,271 @@
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.helix.zookeeper.zkclient.RecursivePersistListener;
+
+
+/**
+ * ZkPathRecursiveWatcherTrie will be used for a registry for ZK persist recursive watcher.
+ * When persist recursive watcher is registered on path /X, ZK will send out data change for any
+ * data/child changing under the tree structure of /X. The event only include the path of changed
+ * ZNode.
+ * In ZkClient, when ever we get a dataChange event for path /X/Y/Z, we need to track back the path
+ * and notify all registered recursive persist listener and notify them about the node change.
+ * ref: https://zookeeper.apache.org/doc/r3.7.1/zookeeperProgrammers.html#sc_WatchPersistentRecursive
+ */
+public class ZkPathRecursiveWatcherTrie {
+
+  /** Root node of PathTrie */
+  private final TrieNode rootNode;
+
+  private final ReadWriteLock lock = new ReentrantReadWriteLock(true);
+
+  private final Lock readLock = lock.readLock();
+
+  private final Lock writeLock = lock.writeLock();
+
+  static class TrieNode {
+
+    final String value;   // Segmented ZNode path at current level
+    final Map<String, TrieNode> children; // A map of segmented ZNode path of next level to TrieNode
+    TrieNode parent;  // A reference to upper level TrieNode
+    Set<RecursivePersistListener> recursiveListeners;
+    // A list of recursive persist watcher on the current path
+
+    /**
+     * Create a trie node with parent as parameter.
+     *
+     * @param parent the parent of this node
+     * @param value the value stored in this node
+     */
+    private TrieNode(TrieNode parent, String value) {
+      this.value = value;
+      this.parent = parent;
+      this.children =
+          new HashMap<>(4);  // We keep the same number of init children side as Zk sevrer
+      this.recursiveListeners = new HashSet<>(4);
+    }
+
+    /**
+     * Get the parent of this node.
+     *
+     * @return the parent node
+     */
+    TrieNode getParent() {
+      return this.parent;
+    }
+
+    /**
+     * set the parent of this node.
+     *
+     * @param parent the parent to set to
+     */
+    void setParent(TrieNode parent) {
+      this.parent = parent;
+    }
+
+    /**
+     * The value stored in this node.
+     *
+     * @return the value stored in this node
+     */
+    public String getValue() {
+      return this.value;
+    }
+
+    /**
+     * Add a child to the existing node.
+     *
+     * @param childName the string name of the child
+     * @param node the node that is the child
+     */
+    void addChild(String childName, TrieNode node) {
+      this.children.putIfAbsent(childName, node);
+    }
+
+    /**
+     * Delete child from this node.
+     *
+     * @param childName the name of the child to be deleted
+     */
+    void deleteChild(String childName) {
```

Comment:
You may need to lock on this? If there is another thread would like to add node.

Reply:
The `TrieNode` class is only accessible with in `ZkPathRecursiveWatcherTrie` class. We do acquire a writ lock when deleting the trie node.